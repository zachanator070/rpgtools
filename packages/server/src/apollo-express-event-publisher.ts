import { EventPublisher } from "./types";
import { PubSub } from "apollo-server-express";
import { injectable } from "inversify";

@injectable()
export class ApolloExpressEventPublisher implements EventPublisher {
	pubsub = new PubSub();

	asyncIterator(events: string[]): AsyncIterator<any> {
		return this.pubsub.asyncIterator(events);
	}

	publish(event: string, payload: any): Promise<void> {
		return this.pubsub.publish(event, payload);
	}
}
