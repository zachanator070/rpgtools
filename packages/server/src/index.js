import express from "express";
import bodyParser from "body-parser";
import gqlServer from "./gql-server";
import cookieParser from "cookie-parser";
import morgan from "morgan";
import mongoose from "mongoose";
import { defaultSeeders } from "./default-seeders";
import path from "path";
import { ImageRouter } from "./routers/image-router";
import { checkConfig, serverNeedsSetup } from "./server-needs-setup";
import crypto from "crypto";
import * as http from "http";
import { ServerConfig } from "./models/server-config";
import { ModelRouter } from "./routers/model-router";
import ExportRouter from "./routers/export-router";
import { graphqlUploadExpress } from "graphql-upload";

const mongodb_host = process.env.MONGODB_HOST || "mongodb";
const mongodb_db_name = process.env.MONGODB_DB_NAME || "rpgtools";

mongoose
  .connect(`mongodb://${mongodb_host}/${mongodb_db_name}`, {
    useNewUrlParser: true,
  })
  .then(async () => {
    console.log(
      `Connected to mongodb at mongodb://${mongodb_host}/${mongodb_db_name}`
    );
    await defaultSeeders();
    await createServer();
  });

mongoose.set("useNewUrlParser", true);
mongoose.set("useFindAndModify", false);
mongoose.set("useCreateIndex", true);
mongoose.set("useUnifiedTopology", true);

const createServer = async () => {
  const server = express();

  server.use(bodyParser.json());
  server.use(morgan("tiny"));

  server.use(cookieParser());

  await checkConfig();
  const serverConfig = await ServerConfig.findOne();
  if (serverNeedsSetup()) {
    console.warn(
      `Server needs configuration! Use unlock code ${serverConfig.unlockCode} to unlock`
    );
  }

  server.get("*.js", function (req, res, next) {
    req.url = req.url + ".gz";
    res.set("Content-Encoding", "gzip");
    res.set("Content-Type", "text/javascript");
    next();
  });
  server.get("*.css", function (req, res, next) {
    req.url = req.url + ".gz";
    res.set("Content-Encoding", "gzip");
    res.set("Content-Type", "text/css");
    next();
  });
  server.get("*favicon.ico", function (req, res, next) {
    req.url = "/favicon.ico";
    next();
  });

  server.use("/images", ImageRouter);
  server.use("/models", ModelRouter);
  server.use("/export", ExportRouter);

  server.get("*", (req, res, next) => {
    const needsSetup = serverNeedsSetup();
    if (needsSetup) {
      if (
        [
          "/api",
          "/ui/setup",
          "/app.bundle.js",
          "/app.css",
          "/app.bundle.js.gz",
          "/app.css.gz",
          "favicon.ico",
        ].includes(req.url)
      ) {
        return next();
      }
      return res.redirect(302, "/ui/setup");
    } else {
      return next();
    }
  });

  server.get("/ui*", (req, res) => {
    return res.sendFile(path.resolve(__dirname, "../../dist", "index.html"));
  });

  server.use(express.static("dist"));
  server.use(express.static("src/static-assets"));
  server.use(graphqlUploadExpress());

  gqlServer.applyMiddleware({ app: server, path: "/api" });

  if (!process.env.ACCESS_TOKEN_SECRET) {
    console.warn(
      "environment variable ACCESS_TOKEN_SECRET is not set, restarting server will log out all users"
    );
    process.env.ACCESS_TOKEN_SECRET = crypto.randomBytes(2048);
  }

  if (!process.env.REFRESH_TOKEN_SECRET) {
    console.log(
      "environment variable REFRESH_TOKEN_SECRET is not set, restarting server will log out all users"
    );
    process.env.REFRESH_TOKEN_SECRET = crypto.randomBytes(2048);
  }

  const port = process.env.SERVER_PORT || 3000;

  const httpServer = http.createServer(server);
  gqlServer.installSubscriptionHandlers(httpServer);

  httpServer.listen(port, () => {
    console.log(
      `The server is running and listening at http://localhost:${port}`
    );
  });
};
