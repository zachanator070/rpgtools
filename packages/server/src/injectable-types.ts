export const INJECTABLE_TYPES = {
	ArticleRepository: Symbol.for("ArticleRepository"),
	ChunkRepository: Symbol.for("ChunkRepository"),
	GameRepository: Symbol.for("GameRepository"),
	ImageRepository: Symbol.for("ImageRepository"),
	ItemRepository: Symbol.for("ItemRepository"),
	ModelRepository: Symbol.for("ModelRepository"),
	MonsterRepository: Symbol.for("MonsterRepository"),
	PermissionAssignmentRepository: Symbol.for("PermissionAssignmentRepository"),
	PersonRepository: Symbol.for("PersonRepository"),
	PinRepository: Symbol.for("PinRepository"),
	PlaceRepository: Symbol.for("PlaceRepository"),
	RoleRepository: Symbol.for("RoleRepository"),
	ServerConfigRepository: Symbol.for("ServerConfigRepository"),
	UserRepository: Symbol.for("UserRepository"),
	WikiFolderRepository: Symbol.for("WikiFolderRepository"),
	WikiPageRepository: Symbol.for("WikiPageRepository"),
	WorldRepository: Symbol.for("WorldRepository"),
	FileRepository: Symbol.for("FileRepository"),

	AuthenticationService: Symbol.for("AuthenticationService"),
	AuthorizationService: Symbol.for("AuthorizationService"),
	ContentExportService: Symbol.for("ContentExportService"),
	ContentImportService: Symbol.for("ContentImportService"),
	GameService: Symbol.for("GameService"),
	ImageService: Symbol.for("ImageService"),
	ModelService: Symbol.for("ModelService"),
	ServerConfigService: Symbol.for("ServerConfigService"),
	SrdImportService: Symbol.for("SrdImportService"),
	UserService: Symbol.for("UserService"),
	WikiFolderService: Symbol.for("WikiFolderService"),
	WikiPageService: Symbol.for("WikiPageService"),
	WorldService: Symbol.for("WorldService"),

	SessionContextFactory: Symbol.for("SessionContextFactory"),
	Open5eApiClient: Symbol.for("Open5eApiClient"),

	Cache: Symbol.for("Cache"),
};
