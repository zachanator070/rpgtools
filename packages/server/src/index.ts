import {container} from "./di/inversify";
import {INJECTABLE_TYPES} from "./di/injectable-types";
import {ApiServer} from "./types";

const server = container.get<ApiServer>(INJECTABLE_TYPES.ApiServer);

(async () => {
	await server.start();
})();
