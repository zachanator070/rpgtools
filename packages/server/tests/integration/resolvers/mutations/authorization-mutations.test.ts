import { WIKI_READ, WIKI_RW, WORLD_READ } from "@rpgtools/common/src/permission-constants";
import { PLACE, WORLD } from "@rpgtools/common/src/type-constants";
import { container } from "../../../../src/di/inversify";
import { RoleFactory, RoleRepository } from "../../../../src/types";
import { INJECTABLE_TYPES } from "../../../../src/di/injectable-types";
import { defaultTestingContextFactory } from "../../DefaultTestingContextFactory";
import {
	ADD_USER_ROLE,
	CREATE_ROLE,
	DELETE_ROLE,
	GRANT_ROLE_PERMISSION,
	GRANT_USER_PERMISSION, REMOVE_USER_ROLE, REVOKE_ROLE_PERMISSION, REVOKE_USER_PERMISSION
} from "@rpgtools/common/src/gql-mutations";
import {AuthorizationService} from "../../../../src/services/authorization-service";

process.env.TEST_SUITE = "authorization-mutations-test";

describe("authorization-mutations", () => {
	let {
		server,
		mockSessionContextFactory,
		otherUser,
		otherUserSecurityContext,
		world,
		testRole,
		currentUser,
		testerSecurityContext,
		newFolder,
		...testingContext
	} = defaultTestingContextFactory();

	const roleRepo = container.get<RoleRepository>(INJECTABLE_TYPES.RoleRepository);
	const authorizationService = container.get<AuthorizationService>(
		INJECTABLE_TYPES.AuthorizationService
	);
	const roleFactory = container.get<RoleFactory>(INJECTABLE_TYPES.RoleFactory);

	describe("with tester logged in and world created", () => {
		beforeEach(async () => {
			({
				mockSessionContextFactory,
				otherUser,
				otherUserSecurityContext,
				world,
				testRole,
				currentUser,
				testerSecurityContext,
				newFolder,
			} = await testingContext.reset());
		});

		test("grantUserPermission", async () => {
			const result = await server.executeGraphQLQuery({
				query: GRANT_USER_PERMISSION,
				variables: {
					userId: otherUser._id.toString(),
					permission: WORLD_READ,
					subjectId: world._id.toString(),
					subjectType: WORLD,
				},
			});
			expect(result).toMatchSnapshot({
				data: {
					grantUserPermission: expect.objectContaining({
						_id: expect.any(String),
						accessControlList: expect.arrayContaining([
							expect.objectContaining({
								_id: expect.any(String),
								permission: expect.any(String),
								canWrite: expect.any(Boolean),
								subject: expect.objectContaining({
									_id: expect.any(String),
									name: expect.any(String),
								}),
							}),
						]),
					}),
				},
				errors: undefined,
			});
		});

		test("grantUserPermission permission denied", async () => {
			mockSessionContextFactory.resetCurrentUser();
			const result = await server.executeGraphQLQuery({
				query: GRANT_USER_PERMISSION,
				variables: {
					userId: otherUser._id.toString(),
					permission: WORLD_READ,
					subjectId: world._id.toString(),
					subjectType: WORLD,
				},
			});
			expect(result).toMatchSnapshot();
		});

		test("revokeUserPermission", async () => {
			await authorizationService.grantUserPermission(
				testerSecurityContext,
				WIKI_READ,
				world.wikiPage,
				PLACE,
				otherUser._id
			);
			const result = await server.executeGraphQLQuery({
				query: REVOKE_USER_PERMISSION,
				variables: {
					userId: otherUser._id.toString(),
					permission: WIKI_READ,
					subjectId: world.wikiPage,
				},
			});
			expect(result).toMatchSnapshot({
				data: {
					revokeUserPermission: {
						_id: expect.any(String),
					},
				},
				errors: undefined,
			});
		});

		test("revokeUserPermission permission denied", async () => {
			mockSessionContextFactory.resetCurrentUser();
			const result = await server.executeGraphQLQuery({
				query: REVOKE_USER_PERMISSION,
				variables: {
					userId: otherUser._id.toString(),
					permission: WORLD_READ,
					subjectId: world._id.toString(),
				},
			});
			expect(result).toMatchSnapshot();
		});

		test("grantRolePermission", async () => {
			const result = await server.executeGraphQLQuery({
				query: GRANT_ROLE_PERMISSION,
				variables: {
					roleId: testRole._id,
					permission: WIKI_READ,
					subjectId: world.wikiPage,
					subjectType: PLACE,
				},
			});
			expect(result).toMatchSnapshot({
				data: {
					grantRolePermission: {
						_id: expect.any(String),
						permissions: [
							{
								_id: expect.any(String),
								subject: {
									_id: expect.any(String),
								},
							},
						],
					},
				},
				errors: undefined,
			});
		});

		test("grantRolePermission permission denied", async () => {
			mockSessionContextFactory.resetCurrentUser();
			const result = await server.executeGraphQLQuery({
				query: GRANT_ROLE_PERMISSION,
				variables: {
					roleId: testRole._id,
					permission: WIKI_READ,
					subjectId: world.wikiPage,
					subjectType: PLACE,
				},
			});
			expect(result).toMatchSnapshot();
		});

		test("revokeRolePermission", async () => {
			await authorizationService.grantRolePermission(
				testerSecurityContext,
				WIKI_READ,
				world.wikiPage,
				PLACE,
				testRole._id
			);
			const result = await server.executeGraphQLQuery({
				query: REVOKE_ROLE_PERMISSION,
				variables: {
					roleId: testRole._id.toString(),
					permission: WIKI_READ,
					subjectId: world.wikiPage,
				},
			});
			expect(result).toMatchSnapshot({
				data: {
					revokeRolePermission: {
						_id: expect.any(String),
						permissions: [
						],
					},
				},
				errors: undefined,
			});
		});

		test("revokeRolePermission permission denied", async () => {
			mockSessionContextFactory.resetCurrentUser();
			await authorizationService.grantRolePermission(
				testerSecurityContext,
				WIKI_READ,
				world.wikiPage,
				PLACE,
				testRole._id
			);
			const result = await server.executeGraphQLQuery({
				query: REVOKE_ROLE_PERMISSION,
				variables: {
					roleId: testRole._id.toString(),
					permission: WIKI_READ,
					subjectId: world.wikiPage,
				},
			});
			expect(result).toMatchSnapshot();
		});

		test("createRole", async () => {
			const result = await server.executeGraphQLQuery({
				query: CREATE_ROLE,
				variables: { worldId: world._id.toString(), name: "new role" },
			});
			expect(result).toMatchSnapshot({
				data: {
					createRole: {
						_id: expect.any(String),
						roles: expect.arrayContaining([
							expect.objectContaining({
								_id: expect.any(String),
								members: expect.arrayContaining([
									expect.objectContaining({
										_id: expect.any(String),
									}),
								]),
								permissions: expect.arrayContaining([
									expect.objectContaining({
										_id: expect.any(String),
										subject: expect.objectContaining({
											_id: expect.any(String),
										}),
									}),
								]),
							}),
						]),
						accessControlList: expect.any(Array),
					},
				},
				errors: undefined,
			});
		});

		test("createRole permission denied", async () => {
			mockSessionContextFactory.setCurrentUser(otherUser);
			const result = await server.executeGraphQLQuery({
				query: CREATE_ROLE,
				variables: { worldId: world._id.toString(), name: "new role" },
			});
			expect(result).toMatchSnapshot();
		});

		test("deleteRole", async () => {
			const role = roleFactory(null, "other delete role", world._id, []);
			await roleRepo.create(role);
			const result = await server.executeGraphQLQuery({
				query: DELETE_ROLE,
				variables: { roleId: role._id.toString() },
			});
			expect(result).toMatchSnapshot({
				data: {
					deleteRole: {
						_id: expect.any(String),
					},
				},
				errors: undefined,
			});
		});

		test("deleteRole permission denied", async () => {
			mockSessionContextFactory.resetCurrentUser();
			const role = roleFactory(null, "other role", world._id, []);
			await roleRepo.create(role);
			const result = await server.executeGraphQLQuery({
				query: DELETE_ROLE,
				variables: { roleId: role._id.toString() },
			});
			expect(result).toMatchSnapshot();
		});

		test("addUserRole", async () => {
			const result = await server.executeGraphQLQuery({
				query: ADD_USER_ROLE,
				variables: {
					userId: otherUser._id.toString(),
					roleId: testRole._id.toString(),
				},
			});
			expect(result).toMatchSnapshot({
				data: {
					addUserRole: {
						_id: expect.any(String),
						roles: expect.arrayContaining([
							expect.objectContaining({
								_id: expect.any(String),
								members: expect.arrayContaining([
									expect.objectContaining({
										_id: expect.any(String),
									}),
								]),
								permissions: expect.arrayContaining([
									expect.objectContaining({
										_id: expect.any(String),
										subject: expect.objectContaining({
											_id: expect.any(String),
										}),
									}),
								]),
								world: {
									_id: expect.any(String),
								},
							}),
						]),
						accessControlList: expect.any(Array),
					},
				},
				errors: undefined,
			});
		});

		test("addUserRole permission denied", async () => {
			mockSessionContextFactory.resetCurrentUser();
			const result = await server.executeGraphQLQuery({
				query: ADD_USER_ROLE,
				variables: {
					userId: otherUser._id.toString(),
					roleId: testRole._id.toString(),
				},
			});
			expect(result).toMatchSnapshot();
		});

		test("removeUserRole", async () => {
			const result = await server.executeGraphQLQuery({
				query: REMOVE_USER_ROLE,
				variables: {
					userId: otherUser._id.toString(),
					roleId: testRole._id.toString(),
				},
			});
			expect(result).toMatchSnapshot({
				data: {
					removeUserRole: {
						_id: expect.any(String),
						roles: expect.arrayContaining([
							expect.objectContaining({
								_id: expect.any(String),
								members: expect.arrayContaining([
									expect.objectContaining({
										_id: expect.any(String),
									}),
								]),
								permissions: expect.arrayContaining([
									expect.objectContaining({
										_id: expect.any(String),
										subject: expect.objectContaining({
											_id: expect.any(String),
										}),
									}),
								]),
								world: {
									_id: expect.any(String),
								},
							}),
						]),
					},
				},
				errors: undefined,
			});
		});

		test("removeUserRole permission denied", async () => {
			await authorizationService.addUserRole(testerSecurityContext, otherUser._id, testRole._id);
			mockSessionContextFactory.resetCurrentUser();
			const result = await server.executeGraphQLQuery({
				query: REMOVE_USER_ROLE,
				variables: {
					userId: otherUser._id.toString(),
					roleId: testRole._id.toString(),
				},
			});
			expect(result).toMatchSnapshot();
		});
	});
});
