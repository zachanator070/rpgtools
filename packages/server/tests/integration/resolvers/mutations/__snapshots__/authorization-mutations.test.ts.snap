// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authorization-mutations addUserRole 1`] = `
Object {
  "data": Object {
    "addUserRole": Object {
      "_id": Any<String>,
      "accessControlList": Any<Array>,
      "canAdmin": true,
      "canWrite": true,
      "roles": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "members": ArrayContaining [
            ObjectContaining {
              "_id": Any<String>,
            },
          ],
          "permissions": ArrayContaining [
            ObjectContaining {
              "_id": Any<String>,
              "subject": ObjectContaining {
                "_id": Any<String>,
              },
            },
          ],
          "world": Object {
            "_id": Any<String>,
          },
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations addUserRole permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to manage this role],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations createRole 1`] = `
Object {
  "data": Object {
    "createRole": Object {
      "_id": Any<String>,
      "accessControlList": Any<Array>,
      "canAdmin": true,
      "canWrite": true,
      "roles": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "members": ArrayContaining [
            ObjectContaining {
              "_id": Any<String>,
            },
          ],
          "permissions": ArrayContaining [
            ObjectContaining {
              "_id": Any<String>,
              "subject": ObjectContaining {
                "_id": Any<String>,
              },
            },
          ],
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations createRole permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to add roles to this world],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations deleteRole 1`] = `
Object {
  "data": Object {
    "deleteRole": Object {
      "_id": Any<String>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations deleteRole permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have write permissions for this role],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations grantRolePermission 1`] = `
Object {
  "data": Object {
    "grantRolePermission": Object {
      "_id": Any<String>,
      "permissions": Array [
        Object {
          "_id": Any<String>,
          "canWrite": true,
          "permission": "Read access to a single wiki page",
          "subject": Object {
            "_id": Any<String>,
            "name": "Earth",
          },
          "subjectType": "Place",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations grantRolePermission permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to assign the permission "Read access to a single wiki page" for this subject],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations grantUserPermission 1`] = `
Object {
  "data": Object {
    "grantUserPermission": ObjectContaining {
      "_id": Any<String>,
      "accessControlList": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "canWrite": Any<Boolean>,
          "permission": Any<String>,
          "subject": ObjectContaining {
            "_id": Any<String>,
            "name": Any<String>,
          },
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations grantUserPermission permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to assign the permission "See this world in search results" for this subject],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations removeUserRole 1`] = `
Object {
  "data": Object {
    "removeUserRole": Object {
      "_id": Any<String>,
      "roles": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "members": ArrayContaining [
            ObjectContaining {
              "_id": Any<String>,
            },
          ],
          "permissions": ArrayContaining [
            ObjectContaining {
              "_id": Any<String>,
              "subject": ObjectContaining {
                "_id": Any<String>,
              },
            },
          ],
          "world": Object {
            "_id": Any<String>,
          },
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations removeUserRole permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to manage this role],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations revokeRolePermission 1`] = `
Object {
  "data": Object {
    "revokeRolePermission": Object {
      "_id": Any<String>,
      "permissions": Array [
        Object {
          "_id": Any<String>,
          "canWrite": true,
          "permission": "Write access to a single wiki page",
          "subject": Object {
            "_id": Any<String>,
            "name": "Earth",
          },
          "subjectType": "Place",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations revokeRolePermission permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to revoke the permission "Read access to a single wiki page"],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations revokeUserPermission 1`] = `
Object {
  "data": Object {
    "revokeUserPermission": Object {
      "_id": Any<String>,
      "accessControlList": Array [],
      "canAdmin": true,
      "canWrite": true,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations revokeUserPermission permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to revoke the permission "See this world in search results"],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created addUserRole 1`] = `
Object {
  "data": Object {
    "addUserRole": Object {
      "_id": Any<String>,
      "accessControlList": Any<Array>,
      "canAdmin": true,
      "canWrite": true,
      "roles": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "members": ArrayContaining [
            ObjectContaining {
              "_id": Any<String>,
            },
          ],
          "permissions": ArrayContaining [
            ObjectContaining {
              "_id": Any<String>,
              "subject": ObjectContaining {
                "_id": Any<String>,
              },
            },
          ],
          "world": Object {
            "_id": Any<String>,
          },
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created addUserRole permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to manage this role],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created createRole 1`] = `
Object {
  "data": Object {
    "createRole": Object {
      "_id": Any<String>,
      "accessControlList": Any<Array>,
      "canAdmin": true,
      "canWrite": true,
      "roles": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "members": ArrayContaining [
            ObjectContaining {
              "_id": Any<String>,
            },
          ],
          "permissions": ArrayContaining [
            ObjectContaining {
              "_id": Any<String>,
              "subject": ObjectContaining {
                "_id": Any<String>,
              },
            },
          ],
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created createRole permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to add roles to this world],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created deleteRole 1`] = `
Object {
  "data": Object {
    "deleteRole": Object {
      "_id": Any<String>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created deleteRole permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have write permissions for this role],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created grantRolePermission 1`] = `
Object {
  "data": Object {
    "grantRolePermission": Object {
      "_id": Any<String>,
      "permissions": Array [
        Object {
          "_id": Any<String>,
          "canWrite": true,
          "permission": "Read access to a single wiki page",
          "subject": Object {
            "_id": Any<String>,
            "name": "Earth",
          },
          "subjectType": "Place",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created grantRolePermission permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to assign the permission "Read access to a single wiki page" for this subject],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created grantUserPermission 1`] = `
Object {
  "data": Object {
    "grantUserPermission": ObjectContaining {
      "_id": Any<String>,
      "accessControlList": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "canWrite": Any<Boolean>,
          "permission": Any<String>,
          "subject": ObjectContaining {
            "_id": Any<String>,
            "name": Any<String>,
          },
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created grantUserPermission permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to assign the permission "See this world in search results" for this subject],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created removeUserRole 1`] = `
Object {
  "data": Object {
    "removeUserRole": Object {
      "_id": Any<String>,
      "roles": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "members": ArrayContaining [
            ObjectContaining {
              "_id": Any<String>,
            },
          ],
          "permissions": ArrayContaining [
            ObjectContaining {
              "_id": Any<String>,
              "subject": ObjectContaining {
                "_id": Any<String>,
              },
            },
          ],
          "world": Object {
            "_id": Any<String>,
          },
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created removeUserRole permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to manage this role],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created revokeRolePermission 1`] = `
Object {
  "data": Object {
    "revokeRolePermission": Object {
      "_id": Any<String>,
      "permissions": Array [
        Object {
          "_id": Any<String>,
          "canWrite": true,
          "permission": "Write access to a single wiki page",
          "subject": Object {
            "_id": Any<String>,
            "name": "Earth",
          },
          "subjectType": "Place",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created revokeRolePermission permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to revoke the permission "Read access to a single wiki page"],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created revokeUserPermission 1`] = `
Object {
  "data": Object {
    "revokeUserPermission": Object {
      "_id": Any<String>,
      "accessControlList": Array [],
      "canAdmin": true,
      "canWrite": true,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`authorization-mutations with tester logged in and world created revokeUserPermission permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission to revoke the permission "See this world in search results"],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
